### PR TITLE
fix(autograd): make linear attention and causal conv checkpoint-safe

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## Before submitting

- [ ] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt). Pending human confirmation; opened as a draft.
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the applicable `AGENTS.md` files.
- [x] I added category, area, and origin labels.

## Affected area

Python autograd integration for linear attention and causal convolution.

## Summary

Read `ctx.saved_tensors` once per backward invocation and slice the resulting tuple in the GDN, GDN2, GDP, KDA, and causal-conv backward bridges. Add regressions exercising `torch.utils.checkpoint.checkpoint(..., use_reentrant=False)` for each operation.

## Why

Non-reentrant checkpointing installs saved-tensor unpack hooks that reject repeated unpacking of the same saved tensors. Repeated access to `ctx.saved_tensors` in one backward invocation raises `CheckpointError`, even though ordinary eager autograd works. Retaining the unpacked tuple within the invocation fixes the checkpoint contract without changing the kernels or retaining it across backward calls.

## API and compatibility impact

No public signature, numerical formula, or kernel scheduling changes. This enables checkpointed training; it is not a latency-speedup claim.

## Testing

Against develop `e45ac96b6d61e50b9dca48eadc97d060fb4c1944`, run the same new tests on the baseline and candidate:

```bash
python -m pytest -q -m 'L0 or L1' \
  test/python/linear_attention/test_la.py \
  test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_semantic_api_sm100.py \
  -k nonreentrant
```

- Baseline with only the new test changes: **5 failed**, all with the expected repeated-unpack `CheckpointError`.
- Candidate: **5 passed**, covering GDN, GDN2, GDP, KDA, and causal conv.
- Local SM100, PyTorch `2.13.0+cu130`, cuDNN backend `9.26`, Python 3.12. Source-isolated runner with an existing compatible compiled binding; no binding changes in this PR.
- Pinned repository pre-commit hooks: passed on all changed files.

Broader architecture and framework CI remains to be run. No new end-to-end or B200 performance measurement is claimed.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved backward-pass reliability for causal convolution and linear-attention operations when using non-reentrant activation checkpointing.
  - Preserved gradient computation for inputs, weights, biases, and other supported parameters.

- **Tests**
  - Added GPU coverage for checkpointed causal-convolution and linear-attention backward passes.
  - Added validation that gradients are produced correctly for all relevant inputs and parameter variants.
  - Strengthened checks for reliable module loading during test execution.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests

**Comment by YangXu1990uiuc** [comment]
I am requesting test CI for `b70ddc4eb9c5` with the following standalone command: `@cudnn-ci-bot run frost oss`. For future updates, request the appropriate test target before review; a maintainer may need to trigger it.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost oss

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `b70ddc4eb9c5` against base `fd409b43d565`.

The fix is sound: each backward invocation unpacks saved tensors once and retains that tuple locally, preserving both checkpoint-reuse branches and optional-input ordering. The earlier GDP support-waiver concern is addressed. No approval-blocking finding; one small CI-coverage follow-up below.

**[P2] Include the new causal-conv checkpoint regression in L0** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/b70ddc4eb9c58e93c886d98b7d5df9e1efdf8bd6/test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_semantic_api_sm100.py#L228))

This regression inherits the module's L1 marker, so the repository's default `-m L0` selection omits the causal-conv half of this fix. The four new linear-attention cases are L0. Please mark this small checkpoint case L0 as well, or explicitly include it in the relevant CI selection; the rest of this L1 file need not move.

Evidence: The new test has no L0 decorator; module-level pytestmark includes L1; test/python/pytest.ini defaults to -m L0. Explicit L0-or-L1 execution reproduced the bug and verified the fix. Attribution: New regression coverage in this PR; the underlying saved-tensor defect predates the PR.

Validation: Independent SM100 RED/GREEN at current develop fd409b43d565 versus b70ddc4eb9c5, using identical new tests: baseline 5/5 fail with the expected checkpoint double-unpack error; head 5/5 pass, zero skips. Covers GDN, GDN2, GDP, KDA and packed causal-conv with bias. Tests collected inside each isolated source tree with explicit -m 'L0 or L1'; verified source paths to avoid the nested-conftest baseline-import trap. PyTorch 2.13/cu130, CuTe DSL 4.8, cuDNN 9.25.1. Native binding sources are unchanged and match the reused extension. Reviewed all five backward changes, both LA checkpoint-reuse branches, and causal-conv optional saved-tensor indexing. No kernel or dispatch changes.

Limitations: Focused SM100 functional validation, not a full multi-architecture regression or performance benchmark. Fresh FROST/OSS CI was requested for this head; it is pending evidence.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
**Codex bot CI follow-up · model: `gpt-6-astra`**

The requested current-head run for `b70ddc4eb9c5` (pipeline 68105004) has finished. All seven FROST lanes passed, including linear attention on SM100 (**1,630 passed, 169 skipped, 8 xfailed**) and SDPA on SM100 (**4,782 passed, 1,806 skipped**).

The overall pipeline is still red from the previously identified BSA/dependency issues: missing `torch` in BSA backward tests, JAX BSA parity running with DSL 4.6.2 although it requires >=4.7, and the API-index job's older JAX installation. Those paths are unchanged by this checkpoint fix. These CI results complement the focused old-head RED/current-head GREEN checkpoint tests in [the review](https://github.com/NVIDIA/cudnn-frontend/pull/958#pullrequestreview-5218480451); the L1 coverage caveat there still applies.

Codex bot review — model `gpt-6-astra`.

**Review by vedaanta — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `b70ddc4eb9c5` against base `9b9f32781d0c`.

The fix preserves saved-tensor ordering and keeps the unpacked tuple local to each backward invocation. I found no P0 or high-risk P1. A fresh run on SM100 after merging with the current target branch passed all five checkpoint regressions, covering GDN, GDN2, GDP, KDA, and causal convolution. The previously reported P2 about default CI selection remains; it is nonblocking, so I am approving.

- **[P2] Still open:** [Include the new causal-conv checkpoint regression in L0](https://github.com/NVIDIA/cudnn-frontend/pull/958#pullrequestreview-5218480451).

Validation: Inspected the complete eight-file diff and traced all five backward bridges, both linear-attention checkpoint-reuse branches, and causal-conv optional saved-input indexing. Only the number of unpack-hook invocations changes; the returned tensors and their order are preserved. Created a clean virtual merge onto current target 9b9f32781d0c and verified that native binding sources are unchanged before reusing a matching extension built against cuDNN 9.26. Ran the five new regressions from the integrated tree with explicit -m 'L0 or L1' and -k nonreentrant on SM100, torch 2.13.0+cu130, CuTeDSL 4.7, and cuDNN 9.26: 5 passed, 4783 unrelated cases deselected, zero skips. Package imports resolved to the isolated integration tree. The prior same-head review independently observed five expected CheckpointError failures on its baseline and five passes on this head. The completed CI pipeline 68105004 has all seven FROST jobs passing; its linear-attention job reports 1630 passed, 169 skipped, and 8 xfailed. Python byte-compilation and git diff --check passed for the integrated changes. GitHub pre-commit passed.

Limitations: The fresh execution was a focused SM100 regression run, not a repeat of the full architecture matrix. No kernel or dispatch changes require a new performance claim. The GitHub merge-requirements check remains red because the PR lacks a milestone and project-board assignment; maintainers need to complete that metadata before merging. The broader OSS pipeline has previously triaged failures in unrelated BSA/dependency paths; it is not an all-green integration result.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Review by jhjpark — APPROVED** [review]
